src/Control/Distributed/Spark/SQL/DataFrame.hs

```diff
@@ -32,3 +33,168 @@ debugDF df = call df "show" []
+                   ('Class "org.apache.spark.sql.DataFrameWriter")
+
+read :: SQLContext -> IO DataFrameReader
+read sc = call sc "read" []
+
+write :: DataFrame -> IO DataFrameWriter
+write df = call df "write" []
+
+readParquet :: [Text] -> DataFrameReader -> IO DataFrame
+readParquet fps dfr = do
+    jfps <- reflect fps
+    call dfr "parquet" [coerce jfps]
+
+writeParquet :: Text -> DataFrameWriter -> IO ()
+writeParquet fp dfw = do
+    jfp <- reflect fp
+    call dfw "parquet" [coerce jfp]
+
+newtype StructType =
+    StructType (J ('Class "org.apache.spark.sql.types.StructType"))
+instance Coercible StructType
+                   ('Class "org.apache.spark.sql.types.StructType")
+
+type JStructFieldClass = 'Class "org.apache.spark.sql.types.StructField"
+
+newtype StructField = StructField (J JStructFieldClass)
+instance Coercible StructField
+                   ('Class "org.apache.spark.sql.types.StructField")
+
+schema :: DataFrame -> IO StructType
+schema df = call df "schema" []
+
+fields :: StructType -> IO [StructField]
+fields st = do
+    jfields <- call st "fields" []
+    Prelude.map StructField <$>
+      reify (jfields :: J ('Array JStructFieldClass))
+
+name :: StructField -> IO Text
+name sf = call sf "name" [] >>= reify
+
+nullable :: StructField -> IO Bool
+nullable sf = call sf "nullable" []
+
+newtype DataType = DataType (J ('Class "org.apache.spark.sql.types.DataType"))
+instance Coercible DataType
+                   ('Class "org.apache.spark.sql.types.DataType")
+
+dataType :: StructField -> IO DataType
+dataType sf = call sf "dataType" []
+
+typeName :: DataType -> IO Text
+typeName dt = call dt "typeName" [] >>= reify
+
+newtype Column = Column (J ('Class "org.apache.spark.sql.Column"))
+instance Coercible Column ('Class "org.apache.spark.sql.Column")
+
+type instance Interp Column = 'Class "org.apache.spark.sql.Column"
+instance Reflect Column ('Class "org.apache.spark.sql.Column") where
+  reflect (Column jcol) = return jcol
+
+select :: DataFrame -> [Column] -> IO DataFrame
+select d1 colexprs = do
+  jcols <- reflect colexprs
+  call d1 "select" [coerce jcols]
+
+filterDF :: DataFrame -> Column -> IO DataFrame
+filterDF d1 colexpr = call d1 "where" [coerce colexpr]
+
+joinOn :: DataFrame -> DataFrame -> Column -> IO DataFrame
+joinOn d1 d2 colexpr = call d1 "join" [coerce d2, coerce colexpr]
+
+unionAll :: DataFrame -> DataFrame -> IO DataFrame
+unionAll d1 d2 = call d1 "unionAll" [coerce d2]
+
+col :: DataFrame -> Text -> IO Column
+col d1 t = do
+  colName <- reflect t
+  call d1 "col" [coerce colName]
+
+-- | Give a 'Column' a new name.
+alias :: Column -> Text -> IO Column
+alias c n = do
+  colName <- reflect n
+  call c "alias" [coerce colName]
+
+lit :: Reflect a ty => a -> IO Column
+lit a =  do
+  c <- upcast <$> reflect a  -- @upcast@ needed to land in java Object
+  callStatic (sing :: Sing "org.apache.spark.sql.functions") "lit" [coerce c]
+
+plus :: Column -> Column -> IO Column
```

Comment:
The column operators that follow, along with the `Column` newtype wrapper, should go in a separate module, called `*.Column`. It can then be imported qualified `as Column`.